contrib/server/operations/pbench-upload-results.py

```diff
@@ -0,0 +1,270 @@
+    """Read the tarball MD5 from the tarball's companion file
+
+    Args:
+        tarball: Path to a tarball file with a {name}.md5 companion
+
+    Returns:
+        The tarball's MD5 value
+    """
+    md5_file = Path(f"{str(tarball)}.md5")
+    return md5_file.read_text().split()[0]
+
+
+def upload(
+    server: str, token: str, tarball: Path, metadata: Optional[list[str]] = None
+) -> Response:
+    query_parameters = {}
+
+    md5 = get_md5(tarball)
+    query_parameters["access"] = "public"
+    dataset = tarball.name
+    satellite, _ = tarball.parent.name.split("::")
+    meta = [
+        f"global.server.legacy.migrated:{datetime.datetime.now(tz=datetime.timezone.utc):%Y-%m-%dT%H:%M}"
+    ]
+    if satellite:
+        meta.append(f"server.origin:{satellite}")
+    if metadata:
+        meta.extend(metadata)
+    query_parameters["metadata"] = meta
+    headers = {
+        "Content-MD5": md5,
+        "content-type": "application/octet-stream",
+        "authorization": f"bearer {token}",
+    }
+
+    with tarball.open("rb") as f:
+        return requests.put(
+            f"{server}/api/v1/upload/{dataset}",
+            headers=headers,
+            params=query_parameters,
+            data=f,
+        )
+
+
+def main() -> int:
+    prog = Path(sys.argv[0])
+    parser = ArgumentParser(prog=prog.name, description="Upload tarballs")
+    parser.add_argument("server", help="Specify the Pbench Server address")
+    parser.add_argument(
+        "tarball", type=Path, help="Specify a tarball or directory path"
+    )
+    parser.add_argument("-b", "--before", help="Select tarballs older than this")
+    parser.add_argument(
+        "-c",
+        "--checkpoint",
+        type=Path,
+        dest="checkpoint",
+        help="Checkpoint file for restart",
+    )
+    parser.add_argument(
+        "-m",
+        "--metadata",
+        action="append",
+        dest="metadata",
+        help="Set metadata on dataset upload",
+    )
+    parser.add_argument(
+        "-p",
+        "--preview",
+        action="store_true",
+        help="Report actions but make no changes",
+    )
+    parser.add_argument(
+        "-s", "--since", action="store", help="Select tarballs no older than this"
+    )
+    parser.add_argument(
+        "-t", "--token", action="store", dest="token", help="Pbench Server API token"
+    )
+    parser.add_argument(
+        "--verify", "-v", dest="verify", action="store_true", help="Show progress"
+    )
+    parsed = parser.parse_args()
+
+    # Get basic configuration
+    host = socket.gethostname()
+    v = Path("/opt/pbench-server/VERSION")
+    s = Path("/opt/pbench-server/SHA1")
+    version = v.read_text().strip() if v.exists() else "unknown"
+    sha1 = s.read_text().strip() if s.exists() else "unknown"
+    metadata = [
+        f"global.server.legacy.hostname:{host}",
+        f"global.server.legacy.sha1:{sha1}",
+        f"global.server.legacy.version:{version}",
+    ]
```

Comment:
I love f-strings for lots of uses, but for simple cases like these they seem like over-kill -- I would use string concatenation or even `":".join()`.  🙂

Reply:
I like the simplicity and readability of f-strings!